source/fundamentals/logging.txt

```diff
@@ -0,0 +1,297 @@
+           <version>1.2.3</version>
+         </dependency>
+
+      Once you have installed the above dependencies, connect to your
+      MongoDB instance and retrieve a document with the following code.
+
+      .. include:: /includes/fundamentals/logging.rst
+
+      You should see output that resembles the following:
+
+      .. code-block:: none
+
+         ...
+         1156 [main] DEBUG o.m.driver.protocol.command - Sending command '{"find": "<your collection>", "filter": {}, "limit": 1, "singleBatch": true, "$db": "<your db>", "lsid": {"id": {"$binary": {"base64": "<server session>", "subType": "04"}}}}' with request id 13 to database <your db> on connection <your connection> to server <your server>
+         1173 [main] DEBUG o.m.driver.protocol.command - Execution of command with request id 13 completed successfully in 30.01 ms on connection <your connection> to server <your server>
+   
+      For more information on Logback, see the
+      `Logback manual <http://logback.qos.ch/manual/>`__.
+   
+   .. tab:: Log4j2
+      :tabid: Log4j2-binding
+   
+      Add the following dependencies to your ``pom.xml`` file.
+
+      .. code-block:: xml      
+      
+         <!-- SLF4J API -->
+         <dependency>
+           <groupId>org.slf4j</groupId>
+           <artifactId>slf4j-api</artifactId>
+           <version>1.7.30</version>
+         </dependency>
+         <!-- SLF4J-Log4j2 binding -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-slf4j-impl</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+         <!-- Log4j2 core functionality -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-core</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+
+      Once you have installed the above dependencies, log an error using the
+      following code.
+
+      .. code-block:: java
+
+         import org.slf4j.Logger;
+         import org.slf4j.LoggerFactory;
+
+         ...
+
+         Logger logger = LoggerFactory.getLogger(App.class);
+         logger.error("Logging an Error");
+
+      You should see output that resembles the following:
+
+      .. code-block:: none
+
+         12:35:00.438 [main] ERROR <my package path> - Logging an Error
+
+      For more information on Log4j2, see the
+      `Log4j2 manual <https://logging.apache.org/log4j/2.x/manual/>`__.
+
+Configure Your Logger
+---------------------
+
+To configure your logger, you must configure the logging framework bound to
+SLF4J.
+
+Example
+~~~~~~~
+
+In this subsection we show how to configure your logger's **log level**.
+
+A logger's log level specifies a lower bound for how urgent a message must be
+for the logger to output that message.
+
+This example shows how to configure your logger's log level to INFO.
+Select the tab corresponding to the logging framework you are using in your
+project.
```

Comment:
We might be able to improve the flow here. 

We state they have to configure the logging framework bound to sfl4j, then we jump into the log level, what it means, and then setting it.

Perhaps in the paragraph under the main heading, we could state the same information and include the info about the log level and tie them together. Then under the example heading, we could state what the example is going to do.

Lastly, is there a cleaner way than "To configure..., you must configure.."? Here's a rough jam of my idea:

```
To configure your logger, set up the logging framework bound to SFL4J by specifying the **log level** and <any other stuff they need to do>.

A logger's log level specifies the lower bound...

Example
~~~~~~~
...
```

Thoughts?

Reply:
Thanks for the note. I agree that "to configure... you must configure..." is not helpful for learners. I have changed it to "to configure.. you must use the configuration system of the logging framework...". This felt less circular. Let me know if this is not satisfactory.

I have move the log level definition into the main heading, and added this transition sentence "In the following example we show how you can use your logging framework's
configuration system to set your logger's **log level**.".

Fix: